File: 45-jump-game-ii/45-jump-game-ii.py
class Solution:
    def jump(self, nums: List[int]) -> int:
        l, r, steps = 0, 0, 0
        while r < len(nums) - 1:
            steps += 1
            new_r = r
            for i in range(l, r + 1):
                if i == len(nums): return steps
                new_r = max(new_r, i + nums[i])
            l, r = r + 1, new_r
        return steps

File: 45-jump-game-ii/45-jump-game-ii.cpp
class Solution {
public:
    int jump(vector<int>& nums) {
        int n = nums.size() - 1, count = 0, currIdx = nums[0], maxIdx = nums[0];
        
        for (int i = 1; i <= n; ++i) {
            if (currIdx < i) {
                ++count;
                currIdx = maxIdx;
            }
            if (currIdx >= n) {
                ++count;
                break;
            }
            
            int nextIdx = i + nums[i];
            if (nextIdx > maxIdx) maxIdx = nextIdx;
        }
        return count;
    }
};